### PR TITLE
feat: Add ability to use custom terraform version

### PR DESCRIPTION
**Pull request by brittandeyoung**
Updated all terraform actions and base "run" action to support specifying a custom version of terraform cli.

### CODE REVIEW

**Review by gerhard — APPROVED** [review]
This looks good from my POV. WDYT @marcosnils @jpadams?

Does it still work as expected for you @brittandeyoung after my changes?

File: pkg/universe.dagger.io/alpha/terraform/run.cue [hunk 1]
```diff
@@ -64,11 +68,12 @@ _#DefaultLogLevel: "off"
 
 	// Hashicorp Terraform container
 	container: #input: docker.#Image | #Image
-
 	// Run command within a container
 	_run: docker.#Build & {
 		steps: [
-			container.#input,
+			container.#input & {
+				"version": version
```
Comment (marcosnils) [inline pkg/universe.dagger.io/alpha/terraform/run.cue:75]: IIUC `container.#input` can be either a `docker.#Image` or `#Image`. 

If a `docker.#Image` is used how come this conjunction could be possible? By looking at this, I'm not sure if the `docker.#Image` flow will kept working after this PR

Comment (brittandeyoung) [reply pkg/universe.dagger.io/alpha/terraform/run.cue]: @marcosnils Good catch! I have moved the input for version to the #Image definition. Please let me know if this resolves your concerns.

**Comment by brittandeyoung** [comment]
Additionally I would like to get thoughts on making the default version `latest`. I bring this up as this may be a great way to keep the Actions up to date with the latest version of the terraform cli. As long as this is documented, it is usually my experience that users will specify the version of terraform they want if they do not want to use `latest`. Otherwise developers will have to specify `latest` or a newer version of the specified default if they need functionality that is in a newer version that the default defined in the `image` action. 

@gerhard @marcosnils

**Comment by marcosnils** [comment]
> @marcosnils Good catch! I have moved the input for version to the `#Image` definition. Please let me know if this resolves your concerns.

thx for updating this :raised_hands:. I just realized that the reason that we didn't catch this in the tests is because we're not currently testing it. WDYT if we add the following action to also test the `docker.#Image` flow? 

```cue
				imageWorkflow: {
			// Set a terraform version for testing specific version of terraform
			_image: docker.#Pull & {
				source: "hashicorp/terraform"
			}

			init: terraform.#Init & {
				container: #input: _image.output
				source: tfSource.output
			}

			workspaceNew: terraform.#Workspace & {
				container: #input: _image.output
				source: init.output
				subCmd: "new"
				name:   "TEST_WORKSPACE"
			}

			plan: terraform.#Plan & {
				container: #input: _image.output
				source: init.output
			}

			apply: terraform.#Apply & {
				container: #input: _image.output
				always: true
				source: plan.output
			}

			verify: #AssertFileRegex & {
				input: apply.output
				path:  "terraform.tfstate"
				regex: string | *"1\\.2\\.7"
			}

			destroy: terraform.#Destroy & {
				container: #input: _image.output
				source: apply.output
				container: #input: _image.output
			}
		}

``` 




> Having this as `latest` would prevent developers from needing to specify version if they are okay with using the latest version and okay with getting auto updates. Then hard coding to a specific version would be more of a conscience effort.

I agree that having `latest` is very convenient for users that prefer to have the latest features without having to hardcode a version in their plans. Having said that, I'd open separate issue to discuss about this since currently a lot of our actions follow the same pattern of having a `_#DefaultVersion` definition and there's a reason that it was done in the first place. If we decide to change this, we should make the effort to make it work equally for all the actions and address it on a single PR if possible.

**Comment by brittandeyoung** [comment]
@marcosnils New test workflow has been added as suggested. I have also opened #2951 to start the discussion of using `latest` for the `_#DefaultVersion`.

**Comment by brittandeyoung** [comment]
The Universe Tests were exceeding the git hub actions timeout value specified, So I increased it.

**Comment by grouville** [comment]
> The Universe Tests were exceeding the git hub actions timeout value specified, So I increased it.

They shouldn't: in general, it times out when bats fails unproperly and a zombie continues to live whenever it shouldn't.

It can happen when a test is broken by the import path. Whenever you test it locally, how long does it take ?

**Comment by brittandeyoung** [comment]
@grouville Unfortunately I am unable to run all the tests locally as I use a M1 Mac. There are tests that rely on docker images that do not have an arm based image. I have been relying on targeting the specific tests I am modifying locally.

**Comment by brittandeyoung** [comment]
@grouville I have seen the Universe test run for close to 23 minutes in past PRs. With tests being added, it would not surprise me if it takes longer than 30 min to run in a PR. This is why I bumped the timeout to test on this PR.

**Comment by gerhard** [comment]
Thank you @brittandeyoung for sticking with this!

I have a few local changes which I am yet to push, but for now it's more important to address the existing comments in reverse order, starting with the most recent one first (bottom to top). Here comes the first one:

> @brittandeyoung I have seen the Universe test run for close to 23 minutes in past PRs. With tests being added, it would not surprise me if it takes longer than 30 min to run in a PR. This is why I bumped the timeout to test on this PR.

This is a symptom of the shared networking available on the free GitHub Actions tier combined with the variable performance of the GitHub Actions Cache.

Locally, these tests consistently take less than 2 minutes to run. Here is my last run:

image

Instead of increasing the timeout, a better fix is to disable the GitHub Actions Caching. This will result in more consistent durations, which currently are around the 15 mins mark. This fix is already in `main`. This has more details: https://github.com/dagger/dagger/pull/2941#issuecomment-1215633072. FTR:

image

My recommendation is to rebase this branch and drop the timeout increase commit. Moving onto the next comment.

**Comment by gerhard** [comment]
> @brittandeyoung Unfortunately I am unable to run all the tests locally as I use a M1 Mac. There are tests that rely on docker images that do not have an arm based image. I have been relying on targeting the specific tests I am modifying locally.

I do not use Docker for Mac since this is 8x slower than the Linux equivalent. FTR: https://docs.dagger.io/1200/local-dev

image

I personally run Docker on Linux & use `DOCKER_HOST=tcp://100.81.87.121:2375` across all my Macs to connect via Tailscale. Technically, I could also use `DOCKER_HOST=tcp://192.168.88.22:2375` since [it's a physical host on the LAN](https://twitter.com/gerhardlazu/status/1509537499567247360), but the Tailscale IP ensures that this works consistently even when I'm remote, on my M1 laptop.

Knowing the 8x slowdown, would it still be worth for you to invest time in getting these tests to work on Docker for Mac ARM?

**Comment by brittandeyoung** [comment]
@gerhard  Thanks for the tips on running it remotely on a linux host. I do not feel these tests need to be "fixed". The examples are the powershell action. The docker image simply does not have a arm64 option. This will likely be the case for a long time. I will work on getting a dev environment using the remote `DOCKER_HOST` as suggested.

**Comment by gerhard** [comment]
> @brittandeyoung Thanks for the tips on running it remotely on a linux host. I will work on getting a dev environment using the remote `DOCKER_HOST` as suggested.

🙌🏻

> @brittandeyoung The examples are the powershell action. The docker image simply does not have a arm64 option. This will likely be the case for a long time. 

The simplest thing that I can think of is to disable those tests if the platform is not `x86_64`. WDYT?

**Comment by brittandeyoung** [comment]
> The simplest thing that I can think of is to disable those tests if the platform is not `x86_64`. WDYT?

I think that might help for tests unrelated. But if the user is modifying this action, I think it would be better for them to get an error when running tests locally then it simply not running them. 

Maybe there should be some documentation indicating potential issues running tests on ARM platforms instead?

**Comment by gerhard** [comment]
It looks like `bats` got "stuck": https://github.com/dagger/dagger/runs/7881156518?check_suite_focus=true#step:6:1097

image

Running this locally now, with an empty cache.

**Comment by gerhard** [comment]
Passes locally on first try:
image

Re-triggered the GitHub Action for now - not holding my breath.

This is my last update in this time-box. Might be able to resume later on today, otherwise it will need to wait for tomorrow.

My takeaway from this session is that my task to provision a dedicated Docker Engine just jumped to the top of my list. These tests should finish in less than 2 minutes. We are wasting a lot of time by not having this. Leaving it here for now. cc @aluzzardi @samalba @shykes

**Comment by brittandeyoung** [comment]
What is interesting when running the tests locally I am also getting this hanging behavior.

image

I get this when running the test against the `main` branch as well. This is not the errors i would get before about missing docker images.

**Comment by gerhard** [comment]
I continued digging into the `make universe-test` timeouts, and this is what I discovered on a fresh Debian 11 host:

image

Notice that we blow past the 4GB limit, start swapping, and then everything slows down to a crawl. While this host is limited to 4GB and the free GitHub Actions Runner has 7GB, I suspect that we are seeing the same issue in GitHub Actions.

> FTR: my local Linux host where this completes in under 2 mins has 64GB of DDR4 and 8 CPUs (16 vCPUs). The fast NVMe drive (7GB/s & 500k IOPS) makes a big difference too.

Next step is to run `make universe-test` in a host limited to 8GB, which is closer to [the GitHub Actions Runner spec]( https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) and see what happens then. My assumption is that the same thing. Will continue tomorrow.

**Comment by gerhard** [comment]
`32287s` (9h) later, the integration tests are still running, and they are unlikely to finish since we are hitting some gRPC timeouts within BuildKit:

image

The above confirms that bumping the timeout to `60m` was unlikely to work.

Before I bump the RAM to 8GB, I am curious what happens if I lower the number of parallel `bats` jobs.

Because we currently run 4 in parallel (meaning 4 `dagger` instances), I suspect that this makes this CUE issue https://github.com/cue-lang/cue/issues/1795 far worse. Based on everything I know, I am fairly certain that we are hitting this specific issue. 👨🏻‍⚕️

**Comment by gerhard** [comment]
Even with `2` bats jobs, we end up with `3` dagger instances which do not fit in 4GB based on the work that they need to do. Notice the RSS taken by dagger processes (ignore the forks) during the test:

Screenshot 2022-08-18 at 11 49 31

And after `make universe-test` is cancelled:

Screenshot 2022-08-18 at 11 50 32

Since the GitHub Actions Runner has 7GB, I am going to bump my test host to 8GB and try again.

**Comment by gerhard** [comment]
Not running multiple universe tests in parallel should keep the memory utilisation within the [free GitHub Actions runner resources](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources).

Each job can use up to 1GB of RAM, and the top-level one that orchestrates all the jobs can use up to 3GB of RAM. This hits the 7GB limit, the runner starts swapping, everything slows down to a crawl, and eventually the `30` mins timeout gets hit. This is what that looks like in one picture:

Screenshot 2022-08-17 at 23 47 38

While the last commit makes the Experimental Universe Tests slower (`9` vs `7` mins), as well as the Integration Tests slower (`11` vs `7` mins), the Universe ones should now be **more reliable** going forward. They also seem to be slightly faster (`14` vs `15` mins). This speed-up might be incidental.

There is a lot more that I discovered since I started looking at this. For now, this improves things enough to get this change merged. I intend to continue digging into this test-related / GitHub Actions issue next week.

Thank you @brittandeyoung for your contribution & sticking with us as we pushed this through. If we did it right, your next contribution should be easier 😎

**Review by gerhard — APPROVED** [review]
👍 🚀

**Pull request merged**